ring/ecm.go

```diff
@@ -0,0 +1,193 @@
+package ring
```

Comment:
Factorization is only used in the `ring` package. Is there a specific reason we would want its interface to be public ? 

We'd better keep it private if it's not the case.

Reply:
In fact, I'd be more inclined to move the factorization to the `utils` package and have it imported by ring.